README.md
# diskover

![diskover](docs/diskover.png?raw=True)

diskover is a filesystem crawler and disk usage analyzer which uses [Amazon Elasticsearch Service](https://aws.amazon.com/elasticsearch-service/) or [Elasticsearch](https://www.elastic.co) and [Kibana](https://www.elastic.co/products/kibana) to index and visualize your data. It is designed to crawl your filesystem and find unneeded files on a local or remote server. Indexed files are bulk added and streamed into Elasticsearch allowing you to visualize the data in Kibana without having to wait until the crawl is finished. diskover aims to be fast, simple and easy to use, and runs in Linux or OS X/macOS. 

### Screenshots

![kibana-screenshot](docs/kibana-dashboard-screenshot.png?raw=True)
![kibana-screenshot](docs/kibana-dashboard-dupes-screenshot.png?raw=True)
![kibana-screenshot](docs/kibana-graph-dupes-screenshot.png?raw=True)
![kibana-screenshot](docs/kibana-graph-hardlinks-screenshot.png?raw=True)


### Installation Guide

#### Download

```sh
git clone https://github.com/shirosaidev/diskover.git
cd diskover
```


### User Guide

#### Requirements

* `Linux or OS X/macOS` (tested on OS X 10.11.6 and Ubuntu 16.04)
* `Python 2.7.` (tested on Python 2.7.10 and 2.7.12)
* `Python client for Elasticsearch`
* `Elasticsearch` (local or AWS ES service, tested on Elasticsearch 5.3.0)
* `Kibana` (tested on Kibana 5.3.0)

#### Optional Install

* [X-Pack](https://www.elastic.co/downloads/x-pack) (for graphs, reports, monitoring and http auth)


#### Getting Started

You need to have at least **Python 2.7.** and have installed **Python client for Elasticsearch** using `pip`:

```sh
python --version
```

```sh
pip install elasticsearch
```

Start diskover as root user with:

```sh
cd /path/you/want/to/crawl
sudo python /path/to/diskover.py
```

A successfull crawl should look like this:

```

   ___       ___       ___       ___       ___       ___       ___       ___
  /\  \     /\  \     /\  \     /\__\     /\  \     /\__\     /\  \     /\  \
 /::\  \   _\:\  \   /::\  \   /:/ _/_   /::\  \   /:/ _/_   /::\  \   /::\  \
/:/\:\__\ /\/::\__\ /\:\:\__\ /::-"\__\ /:/\:\__\ |::L/\__\ /::\:\__\ /::\:\__\
\:\/:/  / \::/\/__/ \:\:\/__/ \;:;-",-" \:\/:/  / |::::/  / \:\:\/  / \;:::/  /
 \::/  /   \:\__\    \::/  /   |:|  |    \::/  /   L;;/__/   \:\/  /   |:\/__/
  \/__/     \/__/     \/__/     \|__|     \/__/    v1.0.7     \/__/     \|__|
                                      https://github.com/shirosaidev/diskover


[2017-05-03 17:58:31] [status] Connecting to Elasticsearch
[2017-05-03 17:58:31] [info] Checking for ES index: diskover-2017.04.22
[2017-05-03 17:58:31] [warning] ES index exists, deleting
[2017-05-03 17:58:31] [info] Creating ES index
Crawling: [100%] |████████████████████████████████████████| 7960/7960
[2017-05-03 17:58:40] [status] Finished crawling
[2017-05-03 17:58:40] [info] Directories Crawled: 7960
[2017-05-03 17:58:40] [info] Files Indexed: 344
[2017-05-03 17:58:40] [info] Elapsed time: 9.41380310059
```


#### diskover CLI arguments

```
usage: diskover.py [-h] [-d TOPDIR] [-m MTIME] [-s MINSIZE] [-t THREADS]
                   [-i INDEX] [-n] [--dupesindex] [-v]

optional arguments:
  -h, --help            show this help message and exit
  -d TOPDIR, --topdir TOPDIR
                        Directory to start crawling from (default: .)
  -m MTIME, --mtime MTIME
                        Minimum days ago for modified time (default: 30)
  -s MINSIZE, --minsize MINSIZE
                        Minimum file size in MB (default: 5)
  -t THREADS, --threads THREADS
                        Number of threads to use (default: 2)
  -i INDEX, --index INDEX
                        Elasticsearch index name (default: from config)
  -n, --nodelete        Do not delete existing index (default: delete index)
  --dupesindex          Create duplicate files index (default: don't create)
  -v, --verbose         Run more verbose (default: not verbose)
```


#### Config file

diskcover will read a local config file `diskover.cfg`. **It needs to be in the same directory as `diskover.py`**.

Here you can exclude any directories and files you don't want to index separated by `,` (**spaces after comma are treated as part of file/directory name**).

Elasticsearch hostname (endpoint), port and index name are also set here. If you are using AWS ES, set `aws = True` and `port = 443` and set the host to the endpoint in your AWS ES console. If you are running Elasticsearch on your localhost or lan, comment out or set `aws = False` and the default port is `9200`. **If you installed Elasticsearch X-Pack, uncomment and set user/password for http auth**.

Lines beginning with `;` are comments and ignored by diskover.

```
[excluded_dirs]
dirs = .snapshot,DO_NOT_DELETE,Databases

[excluded_files]
files = Thumbs.db,.DS_Store,._.DS_Store,.localized,desktop.ini

[elasticsearch]
;aws = False
;host = search-diskover-es-cluster-eg3yztrvzb6qucroyyjk2vokza.ap-northeast-1.es.amazonaws.com
;port = 443
host = localhost
port = 9200
; http auth for X-Pack, uncomment the below two lines if you installed
;user = elastic
;password = changeme
indexname = diskover-2017.04.22
```

#### Speeding up crawl times

diskover **skips empty directories** and will **only index files that are older than `modified time` and larger than `minimum file size`** from the command line options. Excluding certain files/directories will help speed up crawl times as well. **Running with verbose logging will increase crawl times**.

For example, if you wanted to find all the old files that are larger than 10MB that haven't been modified in more than 6 months, you could run diskover with:

```sh
cd /path/you/want/to/crawl
sudo python /path/to/diskover.py -m 180 -s 10
```

You could also speed up the crawl by running multiple `diskover.py` processes and bulk loading into the same `diskover-<name>` index in Elasticsearch. Below is a diagram showing this example.

![diskover-diagram](docs/diskover-diagram.png?raw=True)

#### Benchmarks

Here are some benchmarks running on my Macbook Pro, this includes time to crawl my local filesystem and index in Elasticsearch (in seconds) using the default of 2 threads and single `diskover.py` running.

```
[2017-04-23 10:02:58] [info] Directories Crawled: 10001
[2017-04-23 10:02:58] [info] Files Indexed: 10000
[2017-04-23 10:02:58] [info] Elapsed time: 16.2423961163

[2017-04-23 10:46:24] [info] Directories Crawled: 100001
[2017-04-23 10:46:24] [info] Files Indexed: 100000
[2017-04-23 10:46:24] [info] Elapsed time: 167.5496099
```


### Elasticsearch

#### Indices

diskover creates an index with the name from the config file or from the cli option `-i`. **If an existing index exists with the same name, it will be deleted and a new index created, unless `-n or --nodelete` cli argument is used**. 

If you are doing crawls every week for example, you could name the indices diskover-2017.04.09, diskover-2017.04.16, diskover-2017.04.23, etc. **Index names are required to be `diskover-<string>`**.

#### Append data to existing index

If you are running concurrent `diskover.py` processes you will need to use the `-n or --nodelete` cli argument to append  data to an existing index. See above diagram for example.

#### Duplicate files index

An index for duplicate files can also be created using the `--dupesindex` cli argument **after all crawls are finished**. If you named your index `diskover-2017.05.03`, the index for duplicate files will be named `diskover_dupes-2017.05.03`. See above diagram for example.

#### Generated fields

diskover creates the following fields:

|         Field        |                Description                  |                    Example                  |
|----------------------|---------------------------------------------|---------------------------------------------|
| `last_modified`      | Last modification date                      | `1389220808`                                |
| `last_access`        | Last access date                            | `1482435694`                                |
| `last_change`        | Last change date                            | `1389220808`                                |
| `indexing_date`      | Indexing date                               | `1492612283`                                |
| `filesize`           | File size in bytes                          | `50502536`                                  |
| `filename`           | Original file name                          | `"mypic.png"`                               |
| `extension`          | Original file name extension                | `"png"`                                     |
| `path_parent`        | Parent path name of file                    | `"/tmp/dir1/dir2"`                          |
| `path_full`          | Actual real path name                       | `"/tmp/dir1/dir2/mypic.png"`                |
| `owner`              | Owner name                                  | `"cpark"`                                   |
| `group`              | Group name                                  | `"staff"`                                   |
| `hardlinks`          | Hardlink count                              | `1`                                         |
| `inode`              | Inode number                                | `652490`                                    |
| `filehash`           | MD5 hash of file                            | `3a6949b4b74846a482016d0779560327`          |

#### Filehash

In order to speed up crawl times, a md5 hash for each file is made from combining the strings `filename+filesize+last_modified` and hashing that string, rather than the contents of the file. This seems to be a fast high-level way to create a hash of the file. **You should run the md5 command on the files to compare their hashes before you delete the dupes**.

### Kibana

For the index pattern use `diskover-*`. For the duplicate files use `diskover_dupes-*`. **Make sure the `Index contains time-based events` box is `unchecked`** when you create index patterns.

#### diskover dashboard

To use the diskover dashboard (screenshots above), import the saved objects file `export.json` into Kibana for the dashboard visualizations. In Kibana go to `Management > Saved Objects > Import`.

If nothing is showing in the dashboard, go to `Management > Index Patterns > diskover-*` and then hit the `refresh icon`.

#### Kibana Field Formatting

This will help make the dashboard easier to read like in the screenshot for filesize and dates. In Kibana go to `Management > Index Patterns > diskover-*`. In the `Fields tab` click the `edit icon` under controls column for `filesize` field. Change the format to `bytes` and click `Update Field`. For `access_time, modified_time, and change_time`, edit the fields and change the format to `date` and set the format pattern to `MM-DD-YYYY, HH:mm` and click `Update Field`.

#### Kibana Search Filters

Once you have imported the `export.json` into Kibana, in Kibana's `Discover` page click `open`. There will be a few saved searches in there to help you filter for old files and find duplicate files.

Here are some filter examples:

* `last_modified:[now-5y TO now-3M]` filters files that haven't been modified in over 3 months and less than 5 years
* `last_modified:[now-5y TO now-6M] AND last_access:[now-5y TO now-6M]` filters files that haven't been modified or accessed in over 6 months and less than 5 years
* `last_modified:[now-5y TO now-1y] AND last_access:[now-5y TO now-1y]` filters files that haven't been modified or access in over 1 year and less than 5 years
* `extension:(jpg OR gif OR png OR tif OR tiff OR dpx OR exr OR psd OR bmp OR tga)` filters for image files
* `extension:(aif OR iff OR m3u OR m4a OR mid OR mp3 OR mpa OR wav OR wma)` filters for audio files
* `extension:(asf OR avi OR flv OR m4v OR mov OR mp4 OR mpg OR rm OR vob OR wmv)` filters for video files
* `extension:(cache OR tmp OR temp OR bak OR old)` filters for temp files
* `extension:(7z OR deb OR gz OR pkg OR rar OR rpm OR tar OR zip OR zipx)` filters for compressed files

#### Tracking directory tree size over time

To track a directory tree's change in size over time, create a new index each time crawling from the top-level directory of the project.

Every week or month for example: diskover-PROJA-2017.05.10, diskover-PROJA-2017.05.17, etc.

Then create a new Kibana index pattern `diskover-PROJA-*` to filter to just those indices.

To visualy see the change over time, create a line chart in Kibana using `sum of filesize` for `y-axis` and set `x-axis` to `date-histogram` using `indexing_date` field. Set `interval` to `weekly` or `monthly`.

### X-Pack

#### Graphs

To create the graphs in the screenshots above you need to install [X-Pack](https://www.elastic.co/downloads/x-pack). After X-Pack is installed, edit `diskover.cfg` for http auth credentials since X-Pack adds http auth to Elasticsearch and Kibana.

#### Dupes graph

* index pattern: `diskover_dupes-*`
* verticies field source #1: `filehash` set to `50 max terms per hop`
* verticies field source #2: `inode` set to `50 max terms per hop`
* verticies field source #3: `path_parent` set to `50 max terms per hop`
* settings: uncheck `Significant links` and set `Certainty to 1`
* search filter: `hardlinks: 1`

#### Hardlinks graph

* index pattern: `diskover-*`
* verticies field source #1: `path_parent` set to `50 max terms per hop`
* verticies field source #2: `inode` set to `50 max terms per hop`
* settings: uncheck `Significant links` and set `Certainty to 1`
* search filter: `hardlinks: >1`


## License

```
This software is licensed under the Apache 2 license, quoted below.

Copyright 2017 Chris Park

Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
```


CHANGES
# Diskover Change Log

## [1.0.6] - 2017-05-05
### added
- check for index name is diskover-<string>

## [1.0.5] - 2017-05-04
### fixed
- crash caused from unicode decoding if integer value for owner/group

## [1.0.4] - 2017-05-03
### added
- check for running as root user
- keyboard interrupt
- index field "filehash": md5 hash of file metadata combining filename+filesize+mtime strings
- diskover_dupes-* index created for duplicate files
### changed
- filemeta is now stored in a dictionary instead of a string
- utf-8 decoded all strings stored in filemeta_dict
- file check now checks for symbolic links
- various code cleanup
### fixed
- progress bar printing on new lines if queue was empty
- getting extension from some files caused unicode decode error
### removed
- debug cli option

## [1.0.3] - 2017-04-28
### fixed
- inode field type in the Elasticsearch index mapping from type int to long
- strip new line chars and not spaces from end of directory and file names


diskover.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Diskover fs crawler
# https://github.com/shirosaidev/diskover

import os
import sys
import subprocess
import pwd
import grp
import time
import datetime
import argparse
import Queue
import threading
import ConfigParser
import hashlib
from random import randint
from elasticsearch import Elasticsearch, helpers, RequestsHttpConnection

VERSION = '1.0.9'

def printBanner():
	"""This is the print banner function.
	It prints a random banner.
	"""
	b = randint(1,3)
	# print banner
	if b == 1:
		banner = """
\033[93m
  ________  .__        __
  \______ \ |__| _____|  | _________  __ ___________
   |    |  \|  |/  ___/  |/ /  _ \  \/ // __ \_  __ \\ /)___(\\
   |    `   \  |\___ \|    <  <_> )   /\  ___/|  | \/ (='.'=)
  /_______  /__/____  >__|_ \____/ \_/  \___  >__|   (\\")_(\\")
          \/        \/     \/   v%s      \/
                      https://github.com/shirosaidev/diskover
\033[0m
""" % VERSION
	elif b == 2:
		banner = """
\033[93m
   ___       ___       ___       ___       ___       ___       ___       ___
  /\  \     /\  \     /\  \     /\__\     /\  \     /\__\     /\  \     /\  \\
 /::\  \   _\:\  \   /::\  \   /:/ _/_   /::\  \   /:/ _/_   /::\  \   /::\  \\
/:/\:\__\ /\/::\__\ /\:\:\__\ /::-"\__\ /:/\:\__\ |::L/\__\ /::\:\__\ /::\:\__\\
\:\/:/  / \::/\/__/ \:\:\/__/ \;:;-",-" \:\/:/  / |::::/  / \:\:\/  / \;:::/  /
 \::/  /   \:\__\    \::/  /   |:|  |    \::/  /   L;;/__/   \:\/  /   |:\/__/
  \/__/     \/__/     \/__/     \|__|     \/__/    v%s     \/__/     \|__|
                                      https://github.com/shirosaidev/diskover
\033[0m
""" % VERSION
	elif b == 3:
		banner = """
\033[93m
    _/_/_/    _/            _/
   _/    _/        _/_/_/  _/  _/      _/_/    _/      _/    _/_/    _/  _/_/
  _/    _/  _/  _/_/      _/_/      _/    _/  _/      _/  _/_/_/_/  _/_/
 _/    _/  _/      _/_/  _/  _/    _/    _/    _/  _/    _/        _/
_/_/_/    _/  _/_/_/    _/    _/    _/_/        _/ v%s   _/_/_/  _/
                              https://github.com/shirosaidev/diskover
\033[0m
""" % VERSION
	print(banner)

def printProgressBar(iteration, total, prefix='', suffix=''):
	"""This is the create terminal progress bar function.
	It shows progress of the queue.
	"""
	decimals = 0
	bar_length = 40
	str_format = "{0:." + str(decimals) + "f}"
	percents = str_format.format(100 * (iteration / float(total)))
	filled_length = int(round(bar_length * iteration / float(total)))
	bar = '█' * filled_length + '-' * (bar_length - filled_length)
	sys.stdout.write('\r\033[96m%s [%s%s] |%s| %s\033[0m' % (prefix, percents, '%', bar, suffix)),
	sys.stdout.flush()

def printLog(logtext, logtype='', newline=True):
	"""This is the print log function.
	It prints log ouptput with date and log type prefix.
	"""
	if logtype == 'info':
		 prefix = '[\033[92minfo\033[00m]' # info green
	elif logtype == 'status':
		 prefix = '[\033[93mstatus\033[00m]' # status yellow
	elif logtype == 'warning':
		prefix = '[\033[91mwarning\033[00m]' # warning red
	elif logtype == 'error':
		prefix = '[\033[91merror\033[00m]' # error red
	else:
		prefix = ''
	ts = time.time()
	st = datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S')
	if newline:
		sys.stdout.write('\n')
	sys.stdout.write('\r[%s] %s %s'%(st, prefix, logtext))
	sys.stdout.flush()

def loadConfig():
	"""This is the load config function.
	It checks for config file and loads in
	the config settings.
	"""
	config = ConfigParser.RawConfigParser()
	dir_path = os.path.dirname(os.path.realpath(__file__))
	configfile = '%s/diskover.cfg'% dir_path
	# Check for config file
	if not os.path.isfile(configfile):
		sys.exit('Config file not found')
	config.read(configfile)
	try:
		d = config.get('excluded_dirs', 'dirs')
	except:
		d = ''
		pass
	try:
		f = config.get('excluded_files', 'files')
	except:
		f = ''
		pass
	try:
		AWS = config.get('elasticsearch', 'aws')
	except:
		AWS = False
		pass
	ES_HOST = config.get('elasticsearch', 'host')
	ES_PORT = int(config.get('elasticsearch', 'port'))
	try:
		ES_USER = config.get('elasticsearch', 'user')
	except:
		ES_USER = ''
		pass
	try:
		ES_PASSWORD = config.get('elasticsearch', 'password')
	except:
		ES_PASSWORD = ''
		pass
	INDEXNAME = config.get('elasticsearch', 'indexname')
	EXCLUDED_DIRS = d.split(',')
	EXCLUDED_FILES = f.split(',')

	return AWS, ES_HOST, ES_PORT, ES_USER, ES_PASSWORD, INDEXNAME, EXCLUDED_DIRS, EXCLUDED_FILES

def parseCLIArgs(INDEXNAME):
	"""This is the parse CLI arguments function.
	It parses command line arguments.
	"""
	parser = argparse.ArgumentParser()
	parser.add_argument("-d", "--topdir", default=".", type=str,
						help="Directory to start crawling from (default: .)")
	parser.add_argument("-m", "--mtime", default=30, type=int,
						help="Minimum days ago for modified time (default: 30)")
	parser.add_argument("-s", "--minsize", default=5, type=int,
						help="Minimum file size in MB (default: 5)")
	parser.add_argument("-t", "--threads", default=2, type=int,
						help="Number of threads to use (default: 2)")
	parser.add_argument("-i", "--index", type=str,
						help="Elasticsearch index name (default: from config)")
	parser.add_argument("-n", "--nodelete", action="store_true",
						help="Do not delete existing index (default: delete index)")
	parser.add_argument("--dupesindex", action="store_true",
						help="Create duplicate files index (default: don't create)")
	parser.add_argument("-v", "--verbose", action="store_true",
						help="Run more verbose (default: not verbose)")
	args = parser.parse_args()

	# use index name from command line instead of config file
	if args.index:
		INDEXNAME = args.index
	# check index name
	if INDEXNAME == "diskover" or INDEXNAME.split('-')[0] != "diskover":
		sys.exit('Please name your index: diskover-<string>')

	return args.topdir, args.mtime, args.minsize, args.threads, INDEXNAME, args.nodelete, args.dupesindex, args.verbose

def crawlDirectories(TOPDIR, EXCLUDED_DIRS, DIRECTORY_QUEUE, VERBOSE):
	"""This is the walk directory tree function.
	It crawls the tree top-down using find command.
	Ignores directories that are empty and in
	'EXCLUDED_DIRS'.
	"""
	global total_num_dirs
	cmd = ['find', TOPDIR, '-type', 'd', '-and', '-not', '-empty']
	for i in EXCLUDED_DIRS:
		cmd.append('-and')
		cmd.append('-not')
		cmd.append('-path')
		cmd.append('*%s*' %i)
	printLog('Finding directories to crawl', logtype='status')
	p = subprocess.Popen(cmd,shell=False,stdin=subprocess.PIPE,
					stdout=subprocess.PIPE,stderr=subprocess.PIPE)
	for line in p.stdout:
		# remove any newline chars
		directory = line.rstrip('\r\n')
		if VERBOSE:
			printLog('Queuing directory: %s'%directory, logtype='info')
		# add item to queue (directory)
		DIRECTORY_QUEUE.put(directory)
		total_num_dirs += 1

def crawlFiles(path, DATEEPOCH, DAYSOLD, MINSIZE, EXCLUDED_FILES, VERBOSE):
	"""This is the list directory function.
	It crawls for files using os.listdir.
	Ignores files smaller than 'MINSIZE' MB, newer
	than 'DAYSOLD' old and in 'EXCLUDED_FILES'.
	"""
	global total_num_files
	filelist = []
	# Crawl files in the directory
	for name in os.listdir(path):
		# remove any newline chars
		name = name.rstrip('\r\n')
		# get absolute path (parent of file)
		abspath = os.path.abspath(path)
		# get full path to file
		filename_fullpath = os.path.join(abspath, name)
		# only process if regular file and not symbolic link
		if os.path.isfile(filename_fullpath) and not os.path.islink(filename_fullpath):
			# Skip file if it's excluded
			if name not in EXCLUDED_FILES:
				# Get file modified time
				mtime = int(os.path.getmtime(filename_fullpath))
				# Convert time in days to seconds
				time_sec = DAYSOLD * 86400
				file_mtime_sec = DATEEPOCH - mtime
				# Only process files modified x days ago
				if file_mtime_sec >= time_sec:
					size = int(os.path.getsize(filename_fullpath))
					# Convert bytes to MB
					size_mb = size / 1024 / 1024
					# Skip files smaller than x MB
					if size_mb >= MINSIZE:
						# get file extension
						extension = str(os.path.splitext(filename_fullpath)[1][1:].strip().lower())
						# get access time
						atime = int(os.path.getatime(filename_fullpath))
						# get change time
						ctime = int(os.path.getctime(filename_fullpath))
						# get number of hardlinks
						hardlinks = int(os.stat(filename_fullpath).st_nlink)
						# get inode number
						inode = int(os.stat(filename_fullpath).st_ino)
						# get owner name
						uid = os.stat(filename_fullpath).st_uid
						try:
							owner = pwd.getpwuid(uid).pw_name.split('\\')
							# remove domain before owner
							if len(owner) == 2:
								owner = owner[1]
							else:
								owner = owner[0]
						# if we can't find the owner name, get the uid number
						except KeyError:
							owner = str(uid)
						# get group
						gid = os.stat(filename_fullpath).st_gid
						try:
							group = grp.getgrgid(gid).gr_name.split('\\')
							# remove domain before group
							if len(group) == 2:
								group = group[1]
							else:
								group = group[0]
						# if we can't find the group name, get the gid number
						except KeyError:
							group = str(gid)
						# get time (seconds since epoch)
						indextime = int(time.time())
						# create md5 hash of file using metadata
						filehash = hashlib.md5(name+str(size)+str(mtime)).hexdigest()
						# create file metadata dictionary
						filemeta_dict = {
							"filename": "%s" % name.decode('utf-8'),
							"extension": "%s" % extension.decode('utf-8'),
							"path_full": "%s" % filename_fullpath.decode('utf-8'),
							"path_parent": "%s" % abspath.decode('utf-8'),
							"filesize": size,
							"owner": "%s" % owner.decode('utf-8'),
							"group": "%s" % group.decode('utf-8'),
							"last_modified": mtime,
							"last_access": atime,
							"last_change": ctime,
							"hardlinks": hardlinks,
							"inode": inode,
							"filehash": "%s" % filehash,
							"indexing_date": indextime
							}
						# add file metadata dictionary to filelist list
						filelist.append(filemeta_dict)
						total_num_files += 1
	return filelist

def workerSetup(DIRECTORY_QUEUE, NUM_THREADS, ES, INDEXNAME, DATEEPOCH, DAYSOLD, MINSIZE, EXCLUDED_FILES, VERBOSE):
	threads = []
	for i in range(NUM_THREADS):
		worker = threading.Thread(target=processDirectoryWorker, args=(i, DIRECTORY_QUEUE, ES, INDEXNAME, \
					DATEEPOCH, DAYSOLD, MINSIZE, EXCLUDED_FILES, VERBOSE,))
		worker.setDaemon(True)
		worker.start()
		threads.append(worker)
	return threads

def processDirectoryWorker(threadnum, DIRECTORY_QUEUE, ES, INDEXNAME, DATEEPOCH, DAYSOLD, MINSIZE, EXCLUDED_FILES, VERBOSE):
	"""This is the worker thread function.
	It processes items in the queue one after another.
	These daemon threads go into an infinite loop,
	and only exit when the main thread ends and
	there are no more paths.
	"""
	global total_num_dirs
	filelist = []
	while True:
		if VERBOSE:
			printLog('[%s]: Looking for the next directory'%threadnum, logtype='status')
		# get an item (directory) from the queue
		path = DIRECTORY_QUEUE.get()
		if path is None:
			if not VERBOSE:
				dircount = total_num_dirs - DIRECTORY_QUEUE.qsize()
				printProgressBar(dircount, total_num_dirs, 'Crawling:', '%s/%s'%(dircount,total_num_dirs))
			break
		if VERBOSE:
			printLog('[%s]: Crawling: %s'%(threadnum, path), logtype='info')
		# crawl the files in the directory
		filelist = crawlFiles(path, DATEEPOCH, DAYSOLD, MINSIZE, EXCLUDED_FILES, VERBOSE)
		# add filelist to ES index
		if filelist:
			indexAdd(ES, INDEXNAME, filelist, VERBOSE)
			if not VERBOSE:
				dircount = total_num_dirs - DIRECTORY_QUEUE.qsize()
				printProgressBar(dircount, total_num_dirs, 'Crawling:', '%s/%s'%(dircount,total_num_dirs))
		DIRECTORY_QUEUE.task_done()

def elasticsearchConnect(AWS, ES_HOST, ES_PORT, ES_USER, ES_PASSWORD):
	"""This is the ES function.
	It creates the connection to Elasticsearch
	and checks if it can connect.
	"""
	# Check if we are using AWS ES
	printLog('Connecting to Elasticsearch', logtype='status')
	if AWS:
		ES = Elasticsearch(hosts=[{'host': ES_HOST, 'port': ES_PORT}], \
			use_ssl=True, verify_certs=True, connection_class=RequestsHttpConnection)
	# Local connection to ES
	else:
		ES = Elasticsearch(hosts=[{'host': ES_HOST, 'port': ES_PORT}], \
							http_auth=(ES_USER, ES_PASSWORD))
	# Ping check ES
	if not ES.ping():
		printLog('Unable to connect to Elasticsearch', logtype='error')
		print('\n')
		sys.exit(1)
	return ES

def indexCreate(ES, INDEXNAME, NODELETE):
	"""This is the ES index create function.
	It checks for existing index and deletes if
	there is one with same name. It also creates
	the new index and sets up mappings.
	"""
	printLog('Checking for ES index: %s' % INDEXNAME, logtype='info')
	# check for existing es index
	if ES.indices.exists(index=INDEXNAME):
		# check if nodelete cli argument and don't delete existing index
		if NODELETE:
			printLog('ES index exists, NOT deleting', logtype='info')
			return
		# delete existing index
		else:
			printLog('ES index exists, deleting', logtype='warning')
			ES.indices.delete(index=INDEXNAME, ignore=[400, 404])
	# set up es index mappings and create new index
	mappings = {
		"mappings": {
			"file": {
				"properties": {
					"filename": {
						"type": "keyword"
					},
					"extension": {
						"type": "keyword"
					},
					"path_full": {
						"type": "keyword"
					},
					"path_parent": {
						"type": "keyword"
					},
					"filesize": {
						"type": "long"
					},
					"owner": {
						"type": "keyword"
					},
					"group": {
						"type": "keyword"
					},
					"last_modified": {
						"type": "date",
						"format": "epoch_second"
					},
					"last_access": {
						"type": "date",
						"format": "epoch_second"
					},
					"last_change": {
						"type": "date",
						"format": "epoch_second"
					},
					"hardlinks": {
						"type": "integer"
					},
					"inode": {
						"type": "long"
					},
					"filehash": {
						"type": "keyword"
					},
					"indexing_date": {
						"type": "date",
						"format": "epoch_second"
					}
				}
			}
		}
	}
	printLog('Creating ES index', logtype='info')
	ES.indices.create(index=INDEXNAME, body=mappings)

def indexAdd(ES, INDEXNAME, filelist, VERBOSE):
	"""This is the ES index add function.
	It bulk adds data from worker's crawl
	results into ES.
	"""
	if VERBOSE:
		printLog('Bulk loading to ES index', logtype='info')
	# bulk load data to Elasticsearch index
	helpers.bulk(ES, filelist, index=INDEXNAME, doc_type='file')

def indexCreateDupes(ES, INDEXNAME, NODELETE, VERBOSE):
	"""This is the duplicate file ES index creator.
	It creates a new index of duplicate files from
	an existing index.
	"""
	# search ES for duplicate files
	dupes_list, dupe_count = dupesFinder(ES, INDEXNAME)
	# get index suffix
	INDEXSUFFIX = INDEXNAME.split('diskover-')[1].strip()
	INDEXNAME='diskover_dupes-%s' % INDEXSUFFIX
	# create new index for dupes
	indexCreate(ES, INDEXNAME, NODELETE)
	# add dupes to new index
	indexAdd(ES, INDEXNAME, dupes_list, VERBOSE)
	printLog('Finished creating duplicate files index', logtype='status')
	return dupe_count

def dupesFinder(ES, INDEXNAME):
	"""This is the duplicate file finder function.
	It searches Elasticsearch for files that have the same filehash
	and creates a new index containing those files.
	"""
	dupes_list = []
	dupe_count = 0
	data = {
	"size": 0,
	"aggs": {
	  "duplicateCount": {
	    "terms": {
	    "field": "filehash",
	      "min_doc_count": 2,
		  "size": 1000000
	    },
	    "aggs": {
	      "duplicateDocuments": {
	        "top_hits": {
			  "size": 1000000
			}
	      }
	    }
	  }
	}
	}
	ES.indices.refresh(index=INDEXNAME)
	res = ES.search(index=INDEXNAME, body=data)
	for hit in res['aggregations']['duplicateCount']['buckets']:
		for hit in hit['duplicateDocuments']['hits']['hits']:
			dupes_list.append(hit['_source'])
			dupe_count += 1
	return dupes_list, dupe_count

def main():

	# print random banner
	printBanner()

	if os.geteuid():
		print('\nPlease run as root')
		sys.exit(1)

	# Date calculation seconds since epoch
	DATEEPOCH = int(time.time())

	# load config file
	AWS, ES_HOST, ES_PORT, ES_USER, ES_PASSWORD, INDEXNAME, EXCLUDED_DIRS, EXCLUDED_FILES = loadConfig()

	# parse cli arguments
	TOPDIR, DAYSOLD, MINSIZE, NUM_THREADS, INDEXNAME, NODELETE, DUPESINDEX, VERBOSE = parseCLIArgs(INDEXNAME)

	# connect to Elasticsearch
	ES = elasticsearchConnect(AWS, ES_HOST, ES_PORT, ES_USER, ES_PASSWORD)

	# create duplicate file index if cli argument
	if DUPESINDEX:
		dupe_count = indexCreateDupes(ES, INDEXNAME, NODELETE, VERBOSE)
		# print stats
		elapsedtime = time.time() - DATEEPOCH
		printLog('Found %s duplicates'% dupe_count, logtype='info')
		printLog('Elapsed time: %s'%elapsedtime, logtype='info')
		print('\n')
		sys.exit(0)

	# create Elasticsearch index
	indexCreate(ES, INDEXNAME, NODELETE)

	# Set up directory queue
	DIRECTORY_QUEUE = Queue.Queue()

	# setup worker threads
	THREADS = []
	THREADS = workerSetup(DIRECTORY_QUEUE, NUM_THREADS, ES, INDEXNAME, DATEEPOCH, DAYSOLD, MINSIZE, EXCLUDED_FILES, VERBOSE)

	try:
		# walk directory tree and start crawling
		crawlDirectories(TOPDIR, EXCLUDED_DIRS, DIRECTORY_QUEUE, VERBOSE)

		# wait for all threads to finish
		for i in range(NUM_THREADS):
			DIRECTORY_QUEUE.put(None)
		for t in THREADS:
			t.join()
		printLog('Finished crawling', logtype='status')
		# print stats
		elapsedtime = time.time() - DATEEPOCH
		printLog('Directories Crawled: %s'%total_num_dirs, logtype='info')
		printLog('Files Indexed: %s'%total_num_files, logtype='info')
		printLog('Elapsed time: %s'%elapsedtime, logtype='info')
		print('\n')
		sys.exit(0)

	except KeyboardInterrupt:
		print('\nkeyboard interrupt received, exiting')
		sys.exit(0)

if __name__ == "__main__":
	total_num_files = 0
	total_num_dirs = 0
	main()
